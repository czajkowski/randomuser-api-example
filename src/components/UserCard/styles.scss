@import 'theme';

.container {
    background: #D3D3D3;
    padding: 1rem;
    border-radius: 1rem;
    min-width: 20rem;
    max-width: 26rem;
    margin: 1rem;
    overflow: hidden;
}

.thumbnail {
    height: 5rem;
    width: 5rem;
    margin-right: 1rem;
    margin-bottom: 1rem;
}

.details {
    display: inline-flex;
    flex-direction: column;
    vertical-align: top;
}

.important {
    font-weight: bold;
}

.entry {
    display: flex;
    margin-bottom: 0.2rem;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    max-width: 20rem;

    > :first-child {
        width: 6rem;
        color: #444;
    }
}

.button {
    cursor: pointer;
    margin-top: 1rem;
    width: 10rem;
}