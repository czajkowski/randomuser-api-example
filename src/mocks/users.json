[{
    "id": "1",
    "firstName": "Ryder",
    "lastName": "Pelletier",
    "username": "heavyelephant166",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/men/14.jpg",
    "email": "ryder.pelletier@example.com",
    "phone": "249-324-2721",
    "cell": "932-986-0296",
    "address": {
        "street": "Main St 1342",
        "city": "Inwood",
        "postcode": "V8R 5L1",
        "state": "Alberta"
    }
}, {
    "id": "2",
    "firstName": "Fred",
    "lastName": "Leenhouts",
    "username": "orangefish469",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/men/62.jpg",
    "email": "fred.leenhouts@example.com",
    "phone": "(690)-021-9198",
    "cell": "(282)-272-3009",
    "address": {
        "street": "Elzensteeg 371",
        "city": "Yerseke",
        "postcode": 72166,
        "state": "Overijssel"
    }
}, {
    "id": "3",
    "firstName": "Maurice",
    "lastName": "Prescott",
    "username": "smalldog872",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/men/2.jpg",
    "email": "maurice.prescott@example.com",
    "phone": "(661)-388-6931",
    "cell": "(205)-963-8520",
    "address": {
        "street": "Homestead Rd 4043",
        "city": "Richmond",
        "postcode": 60910,
        "state": "Massachusetts"
    }
}, {
    "id": "4",
    "firstName": "Oline",
    "lastName": "Vedvik",
    "username": "blueswan472",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/women/42.jpg",
    "email": "oline.vedvik@example.com",
    "phone": "27762696",
    "cell": "42938888",
    "address": {
        "street": "Sanatoriebakken 5767",
        "city": "Movik",
        "postcode": "5906",
        "state": "Hordaland"
    }
}, {
    "id": "5",
    "firstName": "Juliette",
    "lastName": "Williams",
    "username": "bluerabbit332",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/women/32.jpg",
    "email": "juliette.williams@example.com",
    "phone": "226-069-7893",
    "cell": "381-083-3073",
    "address": {
        "street": "Lake of Bays Road 6831",
        "city": "Tecumseh",
        "postcode": "L8P 4B9",
        "state": "New Brunswick"
    }
}, {
    "id": "6",
    "firstName": "Buse",
    "lastName": "Barbarosoğlu",
    "username": "purplemouse715",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/women/8.jpg",
    "email": "buse.barbarosoglu@example.com",
    "phone": "(264)-363-3271",
    "cell": "(256)-942-2427",
    "address": {
        "street": "Maçka Cd 6416",
        "city": "Sivas",
        "postcode": 35052,
        "state": "Iğdır"
    }
}, {
    "id": "7",
    "firstName": "Albert",
    "lastName": "Sanchez",
    "username": "sadzebra646",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/men/26.jpg",
    "email": "albert.sanchez@example.com",
    "phone": "00-8814-5606",
    "cell": "0410-111-113",
    "address": {
        "street": "Mcclellan Rd 9849",
        "city": "Tamworth",
        "postcode": 4446,
        "state": "New South Wales"
    }
}, {
    "id": "8",
    "firstName": "Emy",
    "lastName": "Adam",
    "username": "goldengoose960",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/women/57.jpg",
    "email": "emy.adam@example.com",
    "phone": "03-39-32-47-19",
    "cell": "06-26-65-36-65",
    "address": {
        "street": "Rue Abel-Ferry 2997",
        "city": "Saint-Denis",
        "postcode": 31183,
        "state": "Indre-et-Loire"
    }
}, {
    "id": "9",
    "firstName": "باران",
    "lastName": "نجاتی",
    "username": "ticklishfish754",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/women/0.jpg",
    "email": "brn.njty@example.com",
    "phone": "009-80854458",
    "cell": "0939-066-6666",
    "address": {
        "street": "آیت الله مدرس 3857",
        "city": "رشت",
        "postcode": 50980,
        "state": "قم"
    }
}, {
    "id": "10",
    "firstName": "Jonathan",
    "lastName": "Christensen",
    "username": "lazytiger200",
    "thumbnail": "https://randomuser.me/api/portraits/thumb/men/16.jpg",
    "email": "jonathan.christensen@example.com",
    "phone": "52090679",
    "cell": "11069623",
    "address": {
        "street": "Kratvej 2077",
        "city": "Odense Sv",
        "postcode": 99954,
        "state": "Sjælland"
    }
}]