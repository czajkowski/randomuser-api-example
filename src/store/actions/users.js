import { createAction } from 'redux-act';

import randomuserApi from '../../randomuser.api';

export const fetchUsersStart = createAction('Fetch user data - Start');
export const fetchUsersSuccess = createAction('Fetch user data - Success');
export const fetchUsersError = createAction('Fetch user data - Error');

export const fetchUsers = (page) => {
    return (dispatch, getState) => {
        dispatch(fetchUsersStart());

        const {
            users: {
                nationality,
            },
        } = getState();

        return randomuserApi
            .getUsers(page, {
                nationality,
            })
            .then((response) => {
                dispatch(fetchUsersSuccess(response));
            })
            .catch((error) => {
                dispatch(fetchUsersError.asError("Error fetching users!"));
            });
    };
};

export const changeSearch = createAction('Change search');
export const changeNationality = createAction('Change nationality');

export const hideUserDetails = createAction('Hide user details');
export const showUserDetails = createAction('Show user details');
