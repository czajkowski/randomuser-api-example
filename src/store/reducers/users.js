import {
    createReducer
} from 'redux-act';

import {
    fetchUsersStart,
    fetchUsersSuccess,
    fetchUsersError,
    changeSearch,
    changeNationality,
    showUserDetails,
    hideUserDetails,
} from '../actions';

import Nationalities from '../../enums/nationalities.json';

// Get first country code on the list
const DefaultNationality = Object.keys(Nationalities)[0];

const initialState = {
    // Initially I wanted to use a flag in the store to indicate that data are being loaded.
    // This would be useful to show a loader.
    // I've used a component (react-infinite-scroller) that handles that internally.
    // I am leaving the flag in case that needs to be changed.
    isLoading: false,

    // All users returned from API for selected Nationality
    allUsers: [],

    filteredUsers: [],

    // Selected user details to display in modal
    userDetails: null,

    error: null,

    // Last loaded page
    page: -1,

    // All available pages
    totalPages: 0,

    // Selected nationality user to filter users
    nationality: DefaultNationality,

    // Search phrase used to filter users
    search: null,
};

const filterUsers = (users, search) => {
    if (!search) {
        return users;
    }

    const searchString = search.toLowerCase();
    return users
        .filter(({ firstName, lastName }) => `${firstName} ${lastName}`
            .toLowerCase()
            .includes(searchString)
        );
}

export default createReducer({
    [fetchUsersStart]: (state) => ({
        ...state,

        isLoading: true,

        error: null,
    }),

    [fetchUsersSuccess]: (state, { page, totalPages, users }) => ({
        ...state,

        isLoading: false,

        allUsers: [...state.allUsers, ...users],
        filteredUsers: [...state.filteredUsers, ...filterUsers(users, state.search)],

        page,
        totalPages,
    }),

    [fetchUsersError]: (state, error) => ({
        ...state,

        isLoading: false,

        error,
    }),

    [changeSearch]: (state, value) => ({
        ...state,

        // Fund users again in available user list
        filteredUsers: [...filterUsers(state.allUsers, value)],

        search: value,
    }),

    [changeNationality]: (state, value) => ({
        ...state,

        // Clears all data.
        // Starts fetching data all over again.
        allUsers: [],
        filteredUsers: [],

        error: null,

        page: -1,
        totalPages: 0,

        nationality: value,
    }),

    [showUserDetails]: (state, userId) => {
        const user = state.allUsers.find(({ id }) => id === userId);

        return {
            ...state,

            userDetails: user,

            error: user ? null : 'Unable to find user!',
        };
    },

    [hideUserDetails]: (state) => ({
        ...state,

        userDetails: null,
    }),
}, initialState);
