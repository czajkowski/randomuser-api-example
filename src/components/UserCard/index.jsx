import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import styles from './styles.scss';

const UserCard = ({
    id,
    firstName,
    lastName,
    username,
    thumbnail,
    email,

    onClick
}) => (
    <div className={styles.container}>
        <img src={thumbnail} className={styles.thumbnail} />

        <div className={styles.details}>
            <div className={classNames(styles.entry, styles.important)}>
                <div>First name</div>
                <div>{firstName}</div>
            </div>
            <div className={classNames(styles.entry, styles.important)}>
                <div>Last name</div>
                <div>{lastName}</div>
            </div>
            <div className={styles.entry}>
                <div>Username</div>
                <div>{username}</div>
            </div>
            <div className={styles.entry}>
                <div>Email</div>
                <a href="mailto:{email}">{email}</a>
            </div>
            <button className={styles.button} onClick={() => onClick(id)}>
                Details
            </button>
        </div>
    </div>
);

UserCard.propTypes = {
    id: PropTypes.string,
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    username: PropTypes.string,
    thumbnail: PropTypes.string,
    email: PropTypes.string,

    onClick: PropTypes.func
};

export default UserCard;
