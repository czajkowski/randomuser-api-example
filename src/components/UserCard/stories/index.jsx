import React, { useState } from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import UserCard from '../';

import users from '../../../mocks/users.json';

const onClick = action('onClick');

storiesOf('UserCard', module).add('Default', () => (
    <UserCard {...users[0]} onClick={onClick} />
));
