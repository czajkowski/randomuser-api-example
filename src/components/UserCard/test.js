import React from 'react';
import { mount } from 'enzyme';

import UserCard from './';

import users from '../../mocks/users.json';

const testUser = users[0];

describe('UserCard', () => {
    let wrapper;
    let onClick;

    beforeEach(() => {
        onClick = jest.fn();

        wrapper = mount(
            <UserCard
                {...testUser}
                onClick={onClick}
            />
        );
    })

    it('should display basic user data', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should fire callback when button is clicked', () => {
        wrapper.find('button').simulate('click');

        expect(onClick).toHaveBeenCalled();
    })
});
