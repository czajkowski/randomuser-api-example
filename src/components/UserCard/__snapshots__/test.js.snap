// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserCard should display basic user data 1`] = `
<UserCard
  address={
    Object {
      "city": "Inwood",
      "postcode": "V8R 5L1",
      "state": "Alberta",
      "street": "Main St 1342",
    }
  }
  cell="932-986-0296"
  email="ryder.pelletier@example.com"
  firstName="Ryder"
  id="1"
  lastName="Pelletier"
  onClick={[MockFunction]}
  phone="249-324-2721"
  thumbnail="https://randomuser.me/api/portraits/thumb/men/14.jpg"
  username="heavyelephant166"
>
  <div
    className="container"
  >
    <img
      className="thumbnail"
      src="https://randomuser.me/api/portraits/thumb/men/14.jpg"
    />
    <div
      className="details"
    >
      <div
        className="entry important"
      >
        <div>
          First name
        </div>
        <div>
          Ryder
        </div>
      </div>
      <div
        className="entry important"
      >
        <div>
          Last name
        </div>
        <div>
          Pelletier
        </div>
      </div>
      <div
        className="entry"
      >
        <div>
          Username
        </div>
        <div>
          heavyelephant166
        </div>
      </div>
      <div
        className="entry"
      >
        <div>
          Email
        </div>
        <a
          href="mailto:{email}"
        >
          ryder.pelletier@example.com
        </a>
      </div>
      <button
        className="button"
        onClick={[Function]}
      >
        Details
      </button>
    </div>
  </div>
</UserCard>
`;
